util/filesort/filesort.go

```diff
@@ -0,0 +1,392 @@
+	"os"
+	"path"
+	"sort"
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type comparableRow struct {
+	key    []types.Datum
+	val    []types.Datum
+	handle int64
+}
+
+type item struct {
+	index int // source file index
+	value *comparableRow
+}
+
+// Min-heap of comparableRows
+type rowHeap struct {
+	sc     *variable.StatementContext
+	ims    []*item
+	byDesc []bool
+}
+
+func lessThan(sc *variable.StatementContext, i []types.Datum, j []types.Datum, byDesc []bool) bool {
+	for k := range byDesc {
+		v1 := i[k]
+		v2 := j[k]
+
+		ret, err := v1.CompareDatum(sc, v2)
+		if err != nil {
+			panic(err)
+		}
+
+		if byDesc[k] {
+			ret = -ret
+		}
+
+		if ret < 0 {
+			return true
+		} else if ret > 0 {
+			return false
+		}
+	}
+	return false
+}
+
+func (rh *rowHeap) Len() int { return len(rh.ims) }
+
+func (rh *rowHeap) Swap(i, j int) { rh.ims[i], rh.ims[j] = rh.ims[j], rh.ims[i] }
+
+func (rh *rowHeap) Less(i, j int) bool {
+	l := rh.ims[i].value.key
+	r := rh.ims[j].value.key
+	return lessThan(rh.sc, l, r, rh.byDesc)
+}
+
+func (rh *rowHeap) Push(x interface{}) {
+	rh.ims = append(rh.ims, x.(*item))
+}
+
+func (rh *rowHeap) Pop() interface{} {
+	old := rh.ims
+	n := len(old)
+	x := old[n-1]
+	rh.ims = old[0 : n-1]
+	return x
+}
+
+// FileSorter sorts the given rows according to the byDesc order.
+// FileSorter can sort rows that exceed predefined memory capacity.
+type FileSorter struct {
+	keySize   int                        // size of key slice
+	valSize   int                        // size of val slice
+	bufSize   int                        // size of buf slice
+	tmpDir    string                     // working directory for file sort
+	sc        *variable.StatementContext // required by Datum comparison
+	buf       []*comparableRow           // in-momory buffer of rows
+	files     []string                   // files generated by file sort
+	byDesc    []bool                     // whether or not the specific column sorted in descending order
+	fetched   bool
+	rowHeap   *rowHeap
+	fds       []*os.File
+	fileCount int
+}
+
+// NewFileSorter creates a FileSorter instance.
+// Following arguments are mandatory:
+// sc     : StatementContext instance
+//          required by row comparison
+// keySize: the size of the row key
+//          should be positive
+// valSize: the size of the row value
+//          should be positive
+// bufSize: specify the number of rows FileSorter can hold in memory at a time
+//          should be positive
+// byDesc : sepcify the ordering rule of row comparison
+//        : the size of byDesc should match the keySize
+// tmpDir : working directory for FileSorter
+//          caller should make sure differnet FileSorters do not share the same directory
+func NewFileSorter(sc *variable.StatementContext, keySize int, valSize int, bufSize int, byDesc []bool, tmpDir string) (*FileSorter, error) {
```

Comment:
This function has too many arguments, we can create a file sorter builder, set the parameters in a function call, then call `build` to create a `FileSorter`.